Python/Product/PythonTools/PythonTools/Editor/EditorExtensions.cs

```diff
@@ -64,6 +65,59 @@ internal static class EditorExtensions {
             return null;
         }
 
+        internal static SnapshotSpan GetSnapshotSpan(this ITextView textView, LSP.Range range) {
+            Requires.NotNull(range, nameof(range));
+            Requires.NotNull(textView, nameof(textView));
+
+            return textView.TextSnapshot.GetSnapshotSpan(range.Start.Line, range.Start.Character, range.End.Line, range.End.Character);
+        }
+
+        internal static SnapshotSpan GetSnapshotSpan(this ITextSnapshot snapshot, LSP.Range range) {
+            Requires.NotNull(range, nameof(range));
+            Requires.NotNull(snapshot, nameof(snapshot));
+
+            return snapshot.GetSnapshotSpan(range.Start.Line, range.Start.Character, range.End.Line, range.End.Character);
+        }
+
+        internal static SnapshotSpan GetSnapshotSpan(this ITextSnapshot snapshot, int startLine, int? startCharacter, int endLine, int? endCharacter) {
+            if (startLine > endLine || (startLine == endLine && startCharacter > endCharacter)) {
+                return new SnapshotSpan(snapshot, new Span());
+            }
+
+            if (endLine >= snapshot.LineCount) {
+                return new SnapshotSpan(snapshot, new Span());
+            }
+
+            var startSnapshotLine = snapshot.GetLineFromLineNumber(startLine);
+            int startCharacterValue = startCharacter.HasValue ? startCharacter.Value : startSnapshotLine.Length;
+
+            // This should be >, not >=. Lines have one more position than characters, and length gives the number of characters.
+            if (startCharacterValue > startSnapshotLine.Length) {
+                return new SnapshotSpan(snapshot, new Span());
+            }
+
+            var startSnapshotPoint = startSnapshotLine.Start.Add(startCharacterValue);
```

Comment:
do we need to worry about negative numbers?

Reply:
For startCharacterValue? Could do type change to force it to be positive.

Reply:
uint instead of int

Reply:
yes 'startCharacterValue'

Reply:
Actually uint doesn't work because all of the types everywhere use int. If it was negative it would just be lost in a typecast anyway.

I'll add a check instead.